internal/pkg/service/chat.go

```diff
@@ -68,8 +86,76 @@ func (c *Chat) GetActiveUsers() []string {
 	ret := make([]string, 0, len(c.activeUsers))
 
 	for _, v := range c.activeUsers {
-		ret = append(ret, v)
+		ret = append(ret, v.login)
 	}
 
 	return ret
 }
+
+func (c *Chat) ProcessMessage(messageType int, msg string) {
+	log.Printf("ProcessMessage: %s", msg)
+
+}
+
+func (c *Chat) Run(ctx context.Context, messages <-chan string, wsconns <-chan *websocket.Conn) {
+
+	log.Printf("Start working thread")
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		msgStore := []string{} //msgStore is local variable and do not need any synchronization
+		//processing all messages sends in a single thread helps to guarantee proper delivery order
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case msg := <-messages:
+				msgStore = append(msgStore, msg)
+				c.broadcast(msg)
+			case wsconn := <-wsconns:
+				c.sendOldMessages(msgStore, wsconn)
+			}
+		}
+	}()
+}
+
+func (c *Chat) Wait() {
+	log.Printf("Wait working thread to gracefully stop")
+	c.wg.Wait()
+	log.Printf("Working thread gracefully stopped")
+}
+

```

Comment:
broadcast sends message in separate goroutine which may be useful if there are lot of connections, but it still needs to wait until all sends are finished in order to maintain proper delivery order for later messages